### ISSUE TITLE
Add ability to cancel retry and throw exception

### ISSUE THREAD

**Issue by videege** [App-vNext/Polly#67]
Sometimes I want to have a WaitAndRetry policy handling an exception type that will work normally for certain situations but quickly bomb out in other situations, for example:

```
var policy = Policy.Handle<ThirdPartyHttpException>().WaitAndRetryAsync(1, i => TimeSpan.FromSeconds(2*i), async (exception, timespan) => {
    if (exception.StatusCode == 503) {
        Log("failed, trying again...");
    }
    if (exception.StatusCode == 404) {
        throw new ApiException("Not Found"); //this doesn't work
    }
});
```

Basically, I either want a safe way of throwing exceptions from the handler or a way to cancel all future retries so I can fail out quickly when I know something won't succeed.  There isn't a good way to tell based on exception type.  Is there an easy way to do this?

**Comment by reisenberger** [comment]
Hi @videege.  Three thoughts: 

[1] The overload `.Handle<TException>(Func<TException, bool> exceptionPredicate)` allows you to specify a predicate to select in more detail the exceptions you want handled by the policy.  For instance, with:

`var policy = Policy.Handle<ThirdPartyHttpException>(e => e.StatusCode == 503).WaitAndRetryAsync(...)`

... `ThirdPartyHttpException`s with 503 codes only would be handled by the policy and retried; other exceptions would be re-thrown without retries.

[2] In your example, I would expect `throw new ApiException()` within the onRetry delegate to cause further retries to be aborted.  Is this not happening?

[3] We are intending shortly to add cancellation support for the async uses of Polly – see #46 (tho the spec may change slightly).  Various async overloads will be able to take a `CancellationToken`.  Cancelling the related `CancellationTokenSource` would also cause any further retries to be aborted - another avenue to achieving this.

Does this help?

**Comment by reisenberger** [comment]
@videege . Cancellation support has recently been added for async Policy execution - for details, see readme on the Polly home page.  

You can also cancel a related `CancellationTokenSource` within an `onRetry` delegate, and further retries will be cancelled.

In your original question: what about `throw new ApiException("Not Found");` was not working as expected?  Is the scenario, that you are using the given Policy with `.ExecuteAndCaptureAsync()`, and are expecting `.ExecuteAndCaptureAsync()` to capture the ApiException thrown, into the `PolicyResult`?

**Comment by videege** [comment]
Hi, sorry for the delay in replying.  I will try out the new cancellation support.  I got my code working using the approach you suggested by filtering the exception type using the lambda expression available to `Handle`.  I just needed to add in some additional error handling in my wrapper policy for those particular exception types - the normal propagation of exceptions in the handler was what was throwing me off.

Thanks for your help!
